### PR TITLE
feat: Add authentication with tokens

### PR DESCRIPTION
**Pull request by ian-fox**
Related to https://github.com/screwdriver-cd/screwdriver/issues/532

* Token generating function now takes a boolean parameter for if it should put an expiry date on the token
* Token validation function checks that any tokens without an expiry date have not been revoked

### CODE REVIEW

File: plugins/auth/index.js [hunk 1]
```diff
@@ -105,8 +117,25 @@ exports.register = (server, options, next) => {
...
+                        .then(user => user.validateToken(token.uuid))
+                        .then(() => {
+                            cb(null, true, token);
+                        })
+                        .catch(() => {
+                            cb(null, false, {});
+                        });
+                    }
+
+                    // Token is a 12h access token
```
Comment (ian-fox) [inline plugins/auth/index.js]: Notably, if the token is a 12h access token that has expired, it won't even make it to the validateFunc, so if it made it this far we don't need to check that it hasn't expired.

File: plugins/auth/index.js [hunk 2]
```diff
@@ -66,15 +67,26 @@ exports.register = (server, options, next) => {
...
-     * @return {String}         Signed jwt that includes that profile
+     * @param  {Object}     profile     Object from generateProfile
+     * @param  {Boolean}    expires     True if the token should expire
+     * @return {String}                 Signed jwt that includes that profile
      */
-    server.expose('generateToken', profile => jwt.sign(profile, pluginOptions.jwtPrivateKey, {
-        algorithm: ALGORITHM,
-        expiresIn: EXPIRES_IN
-    }));
+    server.expose('generateToken', (profile, expires = true) => {
```
Comment (ian-fox) [inline plugins/auth/index.js]: This is gross and there's probably a better way to do this, please if somebody has an idea let me know.

File: lib/server.js [hunk 3]
```diff
@@ -105,7 +107,8 @@ module.exports = (config) => {
                 server.app.buildFactory.apiUri = server.info.uri;
                 server.app.buildFactory.tokenGen = (buildId, metadata) =>
                     server.plugins.auth.generateToken(
-                        server.plugins.auth.generateProfile(buildId, ['build'], metadata)
+                        server.plugins.auth.generateProfile(buildId, ['build'], metadata),
+                        true
```
Comment (minz1027) [inline lib/server.js]: if you have the default value as `true` then no need for this.

Comment (ian-fox) [inline lib/server.js]: Right you are!

File: plugins/auth/index.js [hunk 4]
```diff
@@ -66,15 +66,23 @@ exports.register = (server, options, next) => {
     });
 
     /**
-     * Generates a jwt that is signed and has a 12h lifespan
+     * Generates a jwt that is signed
      * @method generateToken
-     * @param  {Object} profile Object from generateProfile
-     * @return {String}         Signed jwt that includes that profile
+     * @param  {Object}     profile     Object from generateProfile
+     * @param  {Boolean}    expires     True if the token should expire
```
Comment (minz1027) [inline plugins/auth/index.js]: No need for this.

**Comment by ian-fox** [comment]
Closed as authentication is being redone to trade a different access token for JWT's, rather than just using them directly

**Pull request closed without merging**